[1.4] (2) Fixed FNA discrepancy between Terraria and tModLoader

DarioDaF opened the pull request:
Replaces PR https://github.com/tModLoader/tModLoader/pull/1897 after the recent changes.
Main difference since last fix is the omission of all the logic to neatly organize in subfolders and string manipulation madness done in tML for `.dll`s, figured it wouldn't be a big loss and makes more sense to touch only tML project files if something must change there in the future

DarioDaF committed:
Porting all FNA-dep-solution-fix changes after .NET6 + FNA.Core

Chicken-Bones commented on patches/tModLoader/Terraria/MonoLaunch.cs.patch:
```diff
@@ -1,53 +1,31 @@
  internal static class MonoLaunch
  {
- #if NETCORE
- 	public static readonly object resolverLock = new object();
+@@ -15,7 +_,7 @@
  #endif
-+
-+	private static readonly Dictionary<string, IntPtr> assemblies = new Dictionary<string, IntPtr>();
- 	
+ 
+ 	private static readonly Dictionary<string, IntPtr> assemblies = new Dictionary<string, IntPtr>();
+-
++	
  	private static void Main(string[] args) {
-+		/*
+ 		/*
  		// FNA is requested by both Terraria and ReLogic.dll
- 		var loaded = new Dictionary<string, Assembly>();
- 
+@@ -24,6 +_,7 @@
  		AppDomain.CurrentDomain.AssemblyResolve += delegate (object sender, ResolveEventArgs sargs) {
  			string resourceName = new AssemblyName(sargs.Name).Name + ".dll";
  			string text = Array.Find(typeof(Program).Assembly.GetManifestResourceNames(), (string element) => element.EndsWith(resourceName));
 +			
  			if (text == null)
  				return null;
  
```
Whitespace diffs

Chicken-Bones commented on patches/Terraria/Terraria/Terraria.csproj.patch:
```diff
@@ -74,13 +68,13 @@
 +    </ItemGroup>
 +    <Message Text="Copying $(AssemblyName) to Steam Dir..." Importance="high" />
-+    <Copy SourceFiles="@(BinFiles)" DestinationFolder="$(TerrariaSteamPath)/%(RecursiveDir)" />
++    <Copy SourceFiles="@(BinFiles)" DestinationFolder="$(tModLoaderSteamPath)/%(RecursiveDir)" SkipUnchangedFiles="True" />
```
Intentional change here? We previously built the Terraria sln into the Terraria steam dir. I don't think Terraria.csproj has the required asset path redirects for the side-by-side content sharing etc.

DarioDaF replied:
Fair point I didn't wanna copy the full block there, I should probably just point and do all the stuff tML does, because now a build that is not `Debug` will be called `Terraria.exe` overwriting the game, also a bunch of modified dlls might be used by vanilla terraria due to name clashing (mainly ReLogic).

Chicken-Bones commented on patches/Terraria/Terraria/Terraria.csproj.patch:
```diff
@@ -74,13 +68,13 @@
 +    <ItemGroup>
 +      <BinFiles Include="$(TargetDir)**" />
 +    </ItemGroup>
 +    <Message Text="Copying $(AssemblyName) to Steam Dir..." Importance="high" />
```
Maybe while you're at it, you can put the destination folder in this message

Chicken-Bones reviewed (commented):
Looks mostly good, I'll have to check against my local install that none of the tML files have changed.

DarioDaF committed:
Removed whitespace diff

DarioDaF committed:
Fixing output format and directory

DarioDaF commented:
Ok, moved output directory for Terraria back to the actual Steam folder where `Content` is, moved the code to clean the output to the Terraria solution so it's not a bunch of DLLs potentially conflicting with Vanilla.
Just for cleanliness I added a property with `_` in the solution that is the actual output folder, so both tML and Terraria can use it in the convoluted Targets without diffs

DarioDaF committed:
Correct merge branch '1.4'

Chicken-Bones committed:
Fix a patch offset

Chicken-Bones committed:
Remove a redundant whitespace patch
